autopilot/contractor/contractor.go

```diff
@@ -745,213 +616,171 @@ func renterFundsToExpectedStorage(renterFunds types.Currency, duration uint64, p
-			churnReasons[c.ID] = api.ErrUsabilityHostNotFound.Error()
+			updateUsability(ctx, host, cm, api.ContractUsabilityBad, api.ErrUsabilityHostNotFound.Error())
 			continue
 		}
 
 		// extend logger
-		logger = logger.With("addresses", host.ResolvedAddresses).
+		logger = logger.
+			With("addresses", host.ResolvedAddresses).
 			With("blocked", host.Blocked)
 
 		// check if host is blocked
 		if host.Blocked {
 			logger.Info("host is blocked")
-			churnReasons[c.ID] = api.ErrUsabilityHostBlocked.Error()
+			updateUsability(ctx, host, cm, api.ContractUsabilityBad, api.ErrUsabilityHostBlocked.Error())
 			continue
 		}
 
 		// check if host has a redundant ip
-		if ctx.ShouldFilterRedundantIPs() && ipFilter.HasRedundantIP(host) {
+		if hf.HasRedundantIP(host) {
 			logger.Info("host has redundant IP")
-			churnReasons[c.ID] = api.ErrUsabilityHostRedundantIP.Error()
+			updateUsability(ctx, host, cm, api.ContractUsabilityBad, api.ErrUsabilityHostRedundantIP.Error())
 			continue
 		}
 
 		// get check
-		check, ok := host.Checks[ctx.ApID()]
-		if !ok {
+		if host.Checks == (api.HostChecks{}) {
 			logger.Warn("missing host check")
-			churnReasons[c.ID] = api.ErrUsabilityHostNotFound.Error()
+			updateUsability(ctx, host, cm, api.ContractUsabilityBad, api.ErrUsabilityHostCheckNotFound.Error())
 			continue
 		}
 
-		// NOTE: if we have a contract with a host that is not scanned, we either
-		// added the host and contract manually or reset the host scans. In that case,
-		// we ignore the fact that the host is not scanned for now to avoid churn.
-		if inSet && check.UsabilityBreakdown.NotCompletingScan {
-			keepContract(c.ContractMetadata, host)
+		// NOTE: if we have a contract with a host that is not scanned, we
+		// either added the host and contract manually or reset the host scans.
+		// In that case, we ignore the fact that the host is not scanned for now
+		// to avoid churn.
+		if c.IsGood() && host.Checks.UsabilityBreakdown.NotCompletingScan {
 			logger.Info("ignoring contract with unscanned host")
 			continue // no more checks until host is scanned
 		}
 
 		// check usability
-		if !check.UsabilityBreakdown.IsUsable() {
-			reasons := strings.Join(check.UsabilityBreakdown.UnusableReasons(), ",")
-			logger.With("reasons", reasons).Info("unusable host")
-			churnReasons[c.ID] = reasons
+		if !host.Checks.UsabilityBreakdown.IsUsable() {
+			logger.Info("unusable host")
+			updateUsability(ctx, host, cm, api.ContractUsabilityBad, host.Checks.UsabilityBreakdown.String())
 			continue
 		}
 
 		// check if revision is available
 		if c.Revision == nil {
-			if inSet && remainingLeeway > 0 {
-				logger.Debug("keeping contract due to leeway")
-				keepContract(c.ContractMetadata, host)
-				remainingLeeway--
-			} else {
-				logger.Debug("ignoring contract without revision")
-				churnReasons[c.ID] = errContractNoRevision.Error()
-			}
+			logger.Info("ignoring contract with missing revision")
```

Comment:
@ChrisSchinnerl we skip the contract if we fail to get a revision from the host, thus avoiding unnecessary churn, but at the same time getting rid of the leeway. I wanted to draw attention to it to make sure it's not overlooked. We can also add a leeway but I don't think there's a big downside, while the upside is that we don't have to care about contract sizes when we perform contract checks, which has been a source of many bugs in the past.

Reply:
That's probably fine since the host will sooner or later be offline anyway but you probably still need to call
`updateUsability(ctx, host, cm, c.Usability, host.Checks.UsabilityBreakdown.String())` before `continue` to make sure that if the contract was good, we call `hf.Add(h)`. To make sure we don't end up producing churn through the subnet check.

Reply:
Yeah that was the idea, good catch on the host filter thank you.